### PR TITLE
feat: Add viewmatcher

### PR DESCRIPTION
**Pull request by ki4070ma**
For https://github.com/appium/appium/issues/13747

### CODE REVIEW

**Comment by ki4070ma** [comment]
Need behavior check after below issues resolved.

https://github.com/appium/appium-espresso-driver/issues/503

**Comment by KazuCocoa** [comment]
Did the socket hang error happen?

**Comment by ki4070ma** [comment]
@KazuCocoa 
I'm not sure what you mean is it, but appium server log is here.

appium log
https://gist.github.com/ki4070ma/d955f7159f014a42c0d5996516d782b0

This line shows root cause?
https://gist.github.com/ki4070ma/d955f7159f014a42c0d5996516d782b0#file-gistfile1-txt-L162

**Comment by KazuCocoa** [comment]
Perhaps it was not #503 

What the result of the behaviour of the below adb command on your env?
Then, how about the logcat output if it rases an error?

```
adb shell am instrument -w -e debug false io.appium.espressoserver.test/androidx.test.runner.AndroidJUnitRunner
```

File: appium/webdriver/extensions/search_context/android.py [hunk 1]
```diff
@@ -24,6 +24,66 @@
...
+
+        It works with [Espresso Driver](https://github.com/appium/appium-espresso-driver).
+
+        Args:
+            name (:obj:`str`, optional): The name of a method to invoke.
+                The method must return a Hamcrest
+                [Matcher](http://hamcrest.org/JavaHamcrest/javadoc/1.3/org/hamcrest/Matcher.html)
+            args (:obj:`str`, optional): The args provided to the method
+            className (:obj:`str`, optional): The class name that the method is part of (defaults to `org.hamcrest.Matchers`).
+                Can be fully qualified, or simple, and simple defaults to `androidx.test.espresso.matcher` package
```
Comment (mykola-mokhnach) [inline appium/webdriver/extensions/search_context/android.py]: Can be fully qualified by having the `androidx.test.espresso.matcher.` prefix. If the prefix is not provided then it is going to be added implicitly.

**Comment by ki4070ma** [comment]
@KazuCocoa 
Thanks for your checks.

> What the result of the behaviour of the below adb command on your env?
Not finished below command execution as below.
(Installed io.appium.espressoserver.test package app into the device.)

Can I take any actions?

```
$ adb shell am instrument -w -e debug false io.appium.espressoserver.test/androidx.test.runner.AndroidJUnitRunner

io.appium.espressoserver.EspressoServerRunnerTest:
```

**Comment by KazuCocoa** [comment]
Thanks.

> Can I take any actions?

We should clarify the reason of the below log on your environment. Espresso server looks worked.
Next is maybe making sure if launching `io.appium.android.apis/.ApiDemos` can finish in 45 seconds or something error happens.

> [WD Proxy] Got an unexpected response with status 500: {"id":"f091e8c7-ef39-478e-98d0-9ebf7b7a1401","sessionId":null,"value":{"error":"session not created","message":"java.lang.RuntimeException: Could not launch intent Intent { act=android.intent.action.MAIN flg=0x14000000 cmp=io.appium.android.apis/.ApiDemos } within 45 seconds. Perhaps the main thread has not gone idle within a reasonable amount of time? There could be an animation or something constantly repainting the screen. Or the activity is doing network calls on creation? See the threaddump logs. For your reference the last time the event queue was idle before your activity launch request was 1577868437830 and now the last time the queue went idle was: 1577868443769. If these numbers are the same your activity might be hogging the event queue.","stacktrace":"io.appium.espressoserver.lib.handlers.exceptions.SessionNotCreatedException: java.lang.RuntimeException: Could not launch intent Intent { act=android.intent.action.MAIN flg=0x14000000 cmp=io.appium.android.apis/.ApiDemos } within 45 seconds. Perha...

**Comment by ki4070ma** [comment]
[Notes]
* I have same error log also on Linux.
* I have same error log when trying `npm run e2e-test` with espresso-driver codes on Linux(Ubuntu).
   * https://github.com/appium/appium-espresso-driver#tests

**Comment by ki4070ma** [comment]
@KazuCocoa @mykola-mokhnach 

Espresso driver seems working when using the emulator.

### Questions
Can I ask how to upgrade the espresso driver ver within appium?
I'd like to use ver 1.19.0 or later. (https://github.com/appium/appium-espresso-driver/commits/master)
I've tried to re-install appium and `forceEspressoRebuild` capability, but it seems still using 1.18.0.

Appium log is as below.

[appium log]
```
[debug] [Espresso] 'forceEspressoRebuild' capability is enabled
[debug] [Espresso] Deleting the obsolete Espresso server package '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk'
[Espresso] Building espresso server
[debug] [Espresso] build dir: /var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/espresso-server
[debug] [Espresso] Copying espresso server template from (/usr/local/lib/node_modules/appium/node_modules/appium-espresso-driver/espresso-server to /var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/espresso-server)
[debug] [Espresso] Beginning build with command './gradlew assembleAndroidTest' in directory '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/espresso-server'
[debug] [Espresso] Output from Gradle will not be logged. To change this, use 'showGradleLog' desired capability
[Espresso] Repackaging espresso server for: 'io.appium.android.apis'
[Espresso] Creating new manifest: '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/AndroidManifest.xml'
[debug] [ADB] Found the most recent Android platform: {"platform":"android-28","platformPath":"/Users/atsushimori/Library/Android/sdk/platforms/android-28"}
[ADB] Using 'aapt2' from '/Users/atsushimori/Library/Android/sdk/build-tools/28.0.3/aapt2'
[debug] [ADB] Compiling the manifest using '/Users/atsushimori/Library/Android/sdk/build-tools/28.0.3/aapt2 link -o /var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/AndroidManifest.xml.apk --manifest /var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/AndroidManifest.xml --rename-manifest-package io.appium.espressoserver.test --rename-instrumentation-target-package io.appium.android.apis -I /Users/atsushimori/Library/Android/sdk/platforms/android-28/android.jar -v'
[debug] [ADB] Compiled the manifest at '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/AndroidManifest.xml.apk'
[debug] [ADB] Inserting manifest '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/AndroidManifest.xml', src: '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/espresso-server/app/build/outputs/apk/androidTest/debug/app-debug-androidTest.apk', dst: '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk'
[debug] [ADB] Unzipping '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/AndroidManifest.xml.apk' to '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T'
[debug] [ADB] Unzip successful
[ADB] Using 'aapt' from '/Users/atsushimori/Library/Android/sdk/build-tools/28.0.3/aapt'
[debug] [ADB] Moving manifest
[debug] [ADB] Manifest insertion into '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk' is completed
[Espresso] Repackaged espresso server ready: '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk'
[debug] [ADB] Checking app cert for /var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk
[debug] [ADB] Starting '/Users/atsushimori/Library/Android/sdk/build-tools/28.0.3/apksigner' with args '["verify","--print-certs","/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk"]'
[ADB] Got an error during apksigner execution: Command '/Users/atsushimori/Library/Android/sdk/build-tools/28.0.3/apksigner verify --print-certs /var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk' exited with code 1
[ADB] apksigner stderr: DOES NOT VERIFY
[ADB] ERROR: JAR signer CERT.RSA: JAR signature META-INF/CERT.SF indicates the APK is signed using APK Signature Scheme v2 but no such signature was found. Signature stripped?
[ADB] 
[debug] [ADB] '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk' is not signed with debug cert
[ADB] Using 'zipalign' from '/Users/atsushimori/Library/Android/sdk/build-tools/28.0.3/zipalign'
[debug] [ADB] /var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk' is already zip-aligned. Doing nothing
[debug] [ADB] Signing '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk' with default cert
[debug] [ADB] Starting '/Users/atsushimori/Library/Android/sdk/build-tools/28.0.3/apksigner' with args '["sign","--key","/usr/local/lib/node_modules/appium/node_modules/appium-adb/keys/testkey.pk8","--cert","/usr/local/lib/node_modules/appium/node_modules/appium-adb/keys/testkey.x509.pem","/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk"]'
[debug] [ADB] Uninstalling io.appium.espressoserver.test
[debug] [ADB] Getting install status for io.appium.espressoserver.test
[debug] [ADB] Running '/Users/atsushimori/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell dumpsys package io.appium.espressoserver.test'
[debug] [ADB] 'io.appium.espressoserver.test' is installed
[debug] [ADB] Running '/Users/atsushimori/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell am force-stop io.appium.espressoserver.test'
[debug] [ADB] Running '/Users/atsushimori/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 uninstall io.appium.espressoserver.test'
[debug] [ADB] 'adb uninstall io.appium.espressoserver.test' command output: Success
[ADB] io.appium.espressoserver.test was successfully uninstalled
[Espresso] Uninstalled the obsolete Espresso server package from the device under test
[debug] [ADB] Getting install status for io.appium.espressoserver.test
[debug] [ADB] Running '/Users/atsushimori/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell dumpsys package io.appium.espressoserver.test'
[debug] [ADB] 'io.appium.espressoserver.test' is not installed
[debug] [ADB] App '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk' is not installed
[Espresso] Installing Espresso Test Server apk from the target device (path: '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk')
[debug] [ADB] Running '/Users/atsushimori/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell 'ls -t -1 /data/local/tmp/appium_cache 2>&1 || echo _ERROR_''
[debug] [ADB] The count of applications in the cache: 5
[ADB] Caching the application at '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk' to '/data/local/tmp/appium_cache/7a258aa389e555a7482e64e0be34705d6ed45f77.apk'
[debug] [ADB] Running '/Users/atsushimori/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell mkdir -p /data/local/tmp/appium_cache'
[debug] [ADB] Running '/Users/atsushimori/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 push /var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk /data/local/tmp/appium_cache/7a258aa389e555a7482e64e0be34705d6ed45f77.apk'
[ADB] The upload of 'io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk' (4.77 MB) took 0.336s
[debug] [ADB] Running '/Users/atsushimori/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell pm install /data/local/tmp/appium_cache/7a258aa389e555a7482e64e0be34705d6ed45f77.apk'
[ADB] The installation of 'io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk' took 8.964s
[debug] [ADB] Install command stdout: Success
[Espresso] Installed Espresso Test Server apk '/var/folders/vp/7r1bnmd91zq18mb6qz8zkgtc0000gn/T/io.appium.espressoserver.test_1.18.0_io.appium.android.apis.apk' (pkg: 'io.appium.espressoserver.test')
[debug] [Espresso] Performing cleanup of automation leftovers
[debug] [Espresso] No obsolete sessions have been detected (Error: socket hang up)
[Espresso] Starting Espresso Server v1.18.0 with cmd: adb shell am instrument -w -e debug false io.appium.espressoserver.test/androidx.test.runner.AndroidJUnitRunner
[debug] [ADB] Creating ADB subprocess with args: ["-P",5037,"-s","emulator-5554","shell","am","instrument","-w","-e","debug","false","io.appium.espressoserver.test/androidx.test.runner.AndroidJUnitRunner"]
[debug] [Espresso] [Instrumentation] [STDOUT]
[Espresso] Waiting for Espresso to be online...
[debug] [WD Proxy] Matched '/status' to command name 'getStatus'
[debug] [WD Proxy] Proxying [GET /status] to [GET http://localhost:8300/status] with no body

```

**Comment by mykola-mokhnach** [comment]
@ki4070ma Did you install appium@beta?

**Comment by ki4070ma** [comment]
@mykola-mokhnach 
Thanks for your quick response, I haven't checked below file.
Now `viewmatcher` works with appium v1.16.0-beta3.

v1.16.0 -> `"appium-espresso-driver": "^1.18.0",`
https://github.com/appium/appium/blob/v1.16.0/package.json#L47

v1.16.0-beta3 -> `"appium-espresso-driver": "^1.0.0"`
https://github.com/appium/appium/blob/v1.16.0-beta.3/package.json#L47

File: test/functional/android/search_context/find_by_view_matcher_tests.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            args={
+                'name': 'containsString',
+                'args': 'Animati',
+                'class': 'org.hamcrest.Matchers'},
+            className='ViewMatchers')
+        assert el.text == 'Animation'
+
+    # androidx.test.espresso.AmbiguousViewMatcherException:
+    # 'with text: a string containing "Access"' matches multiple views in the hierarchy.
+    def test_find_multiple_elements(self):
```
Comment (ki4070ma) [inline test/functional/android/search_context/find_by_view_matcher_tests.py]: This is reason not to have `find_elements_by_android_view_matcher`.

**Pull request merged**